harness/determined/core/_checkpoint.py

```diff
@@ -34,6 +34,143 @@ class DownloadMode(enum.Enum):
     NoSharedDownload = "NO_SHARED_DOWNLOAD"
 
 
+def _merge_metadata(
+    merged: Dict[str, Any],
+    rank_metadata: Dict[str, Any],
+    rank: int,
+    key_ranks: Dict[str, Any],
+    key_conflicts: Dict[str, List[int]],
+    prev_key: str,
+) -> Tuple[Dict[str, Any], Dict[str, Any]]:
+
+    for key, metadata in rank_metadata.items():
+
+        # First, update key_ranks. We will use it later
+        if key not in key_ranks:
+            key_ranks[key] = {"_ranks_": []}
+        key_ranks[key]["_ranks_"].append(rank)
+
+        full_key = f"{prev_key}/{key}"
+
+        # Merge metadata. We consider two possibilities:
+        # 1. metadata is dictionary AND :
+        #   * if key is not in merged, then recursively unroll and merge metadata; OR
+        #   * if key is in merged, and metadata and merged[key] are dictionaries, merge recursively.
+        # 2. otherwise:
+        #   * key is in merged, then updated key_conflicts and don't merge,
+        #   * key is not in merged, then add key and metadata to merged.
+        if (key not in merged and isinstance(metadata, dict)) or (
+            key in merged and isinstance(metadata, dict) and isinstance(merged[key], dict)
+        ):
+            merged[key], key_ranks[key] = _merge_metadata(
+                {} if key not in merged else merged[key],
+                metadata,
+                rank,
+                key_ranks[key],
+                key_conflicts,
+                full_key,
+            )
+        else:
+            if key in merged:
+                if full_key not in key_conflicts:
+                    key_conflicts[full_key] = key_ranks[key]["_ranks_"].copy()
+                else:
+                    key_conflicts[full_key].append(rank)
+            else:
+                merged[key] = metadata
+
+    return merged, key_ranks
+
+
+def merge_metadata(
+    all_metadata: List[Dict[str, Any]]
+) -> Tuple[Dict[str, Any], Dict[str, List[int]]]:
+    """
+    Given a list of metadata, return:
+      - merged metadata
+      - a dict mapping conflicting keys to ranks that would upload them
+
+    Merging scenarios:
+      - merge dictionaries under the same key
+      - report conflicts for any repeated key that is not a dictionary
+
+    Examples:
+    #  Metadata 1     | Metadata 2       | Conflict |  Result
+    # ----------------|------------------|----------|-----------------
+    #  a: 1           | b: 2             | no       | {a: 1, b: 2}
+    #  a: 1           | a: 1             | yes      | n/a
```

Comment:
can we allow duplicated keys as long as values are the same across ranks?  without allowing it, the user has to write code that toggles the metadata based on rank for `steps_completed` which we require in metadata when reporting a checkpoint.  this is pretty annoying and not intuitive.